src/main/java/com/microsoft/aad/msal4j/HttpHelper.java

```diff
@@ -42,17 +73,79 @@ static IHttpResponse executeHttpRequest(final HttpRequest httpRequest,
                 HttpHelper.verifyReturnedCorrelationId(httpRequest, httpResponse);
             }
         }
+        processThrottlingInstructions(httpResponse, requestContext);
+
+        return httpResponse;
+    }
+
+    private static IHttpResponse executeHttpRequestWithRetries(HttpRequest httpRequest, IHttpClient httpClient)
+            throws Exception {
+        IHttpResponse httpResponse = null;
+        for (int i = 0; i < RETRY_NUM; i++) {
+            httpResponse = httpClient.send(httpRequest);
+            if (!isRetriable(httpResponse)) {
+                break;
+            }
+            Thread.sleep(RETRY_DELAY_MS);
+        }
         return httpResponse;
     }
 
-    private static void addRequestInfoToTelemetry(final HttpRequest httpRequest, HttpEvent httpEvent){
-        try{
+    private static void checkForThrottling(RequestContext requestContext) {
+        if (requestContext.clientApplication() instanceof PublicClientApplication &&
+                requestContext.apiParameters() != null) {
+            String requestThumbprint = getRequestThumbprint(requestContext);
+
+            long retryInMs = ThrottlingCache.retryInMs(requestThumbprint);
+
+            if (retryInMs > 0) {
+                throw new MsalThrottlingException(retryInMs);
+            }
+        }
+    }
+
+    private static void processThrottlingInstructions(IHttpResponse httpResponse, RequestContext requestContext) {
+        if (requestContext.clientApplication() instanceof PublicClientApplication) {
+            Long expirationTimestamp = null;
+
+            Integer retryAfterHeaderVal = getRetryAfterHeader(httpResponse);
+            if (retryAfterHeaderVal != null) {
+                expirationTimestamp = System.currentTimeMillis() + retryAfterHeaderVal * 1000;
+            } else if (httpResponse.statusCode() == 429 ||
+                    (httpResponse.statusCode() >= 500 && httpResponse.statusCode() < 600)) {
+
+                expirationTimestamp = System.currentTimeMillis() + ThrottlingCache.DEFAULT_THROTTLING_TIME_SEC * 1000;
+            }
+            if (expirationTimestamp != null) {
+                ThrottlingCache.set(getRequestThumbprint(requestContext), expirationTimestamp);
+            }
+        }
+    }
+
+    private static Integer getRetryAfterHeader(IHttpResponse httpResponse) {
+
+        if (httpResponse.headers() != null) {
+            TreeMap<String, List<String>> headers = new TreeMap<>(String.CASE_INSENSITIVE_ORDER);
+            headers.putAll(httpResponse.headers());
+
+            if (headers.containsKey(RETRY_AFTER_HEADER) && headers.get(RETRY_AFTER_HEADER).size() == 1) {
+                int headerValue = Integer.parseInt(headers.get(RETRY_AFTER_HEADER).get(0));
+                if (headerValue > 0 && headerValue <= ThrottlingCache.MAX_THROTTLING_TIME_SEC) {
```

Comment:
if headerValue > MAX or cannot be parsed, perhaps we should behave as if no "RetryAfter" header was set, i.e. use DEFAULT_THROTTLING_TIME_SEC instead of NULL, at least if response is 429 or 5xx.

Reply:
"we should behave as if no "RetryAfter" header was set, i.e. use DEFAULT_THROTTLING_TIME_SEC, at least if response is 429 or 5xx." - that is exactly what happens

Reply:
Yes, I must have read the code wrong.